### PR TITLE
Some tests for networking and SSL

### PR DESCRIPTION
**Pull request by ajbt200128**
The ocaml ssl situation is... bad. Here are some tests to help

### CODE REVIEW

**Comment by ajbt200128** [comment]
for reference, here are the local errors:
```
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ [FAIL]        OSemgrep Networking > Basic HTML                                     1   GET asynchronously.                                                                                                                                               │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
[00.77][DEBUG]: GET asynchronously (parallel)
[00.77][DEBUG]: (lwt) GET https://www.google.com/
[exception] TLS failure: signature verification failed: bad signature
            Raised at Http_helpers.get_async.(fun) in file "libs/networking/http_helpers.ml", line 24, characters 2-692
            Re-raised at Unit_Networking.get_and_check_lwt.(fun) in file "src/osemgrep/networking/Unit_Networking.ml", line 37, characters 2-191
            Re-raised at Unit_Networking.get_and_check_multi_lwt.(fun) in file "src/osemgrep/networking/Unit_Networking.ml", line 53, characters 11-81
            Re-raised at Lwt.Miscellaneous.poll in file "src/core/lwt.ml", line 3123, characters 20-29
            Called from Lwt_main.run.run_loop in file "src/unix/lwt_main.ml", line 27, characters 10-20
            Called from Lwt_main.run in file "src/unix/lwt_main.ml", line 106, characters 8-13
            Re-raised at Lwt_main.run in file "src/unix/lwt_main.ml", line 112, characters 4-13
            Called from Alcotest_engine__Core.Make.protect_test.(fun) in file "src/alcotest-engine/core.ml", line 181, characters 17-23
            Called from Alcotest_engine__Monad.Identity.catch in file "src/alcotest-engine/monad.ml", line 24, characters 31-35
            
Logs saved to `~/r2c/semgrep/_build/_tests/semgrep-core/OSemgrep Networking U+003E Basic HTML.001.output'.
 ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────


┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ [FAIL]        OSemgrep Networking > Basic HTML                                     2   GET asynchronously (parallel).                                                                                                                                    │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
[00.91][DEBUG]: GET asynchronously (parallel)
[00.91][DEBUG]: (lwt) GET https://www.google.com/
[00.91][DEBUG]: (lwt) GET https://semgrep.dev/
[00.91][DEBUG]: (lwt) GET https://github.com/
ASSERT Body is not empty
[exception] TLS failure: authentication failure: invalid certificate chain
            Raised at Lwt.Miscellaneous.poll in file "src/core/lwt.ml", line 3123, characters 20-29
            Called from Lwt_main.run.run_loop in file "src/unix/lwt_main.ml", line 27, characters 10-20
            Called from Lwt_main.run in file "src/unix/lwt_main.ml", line 106, characters 8-13
            Re-raised at Lwt_main.run in file "src/unix/lwt_main.ml", line 112, characters 4-13
            Called from Alcotest_engine__Core.Make.protect_test.(fun) in file "src/alcotest-engine/core.ml", line 181, characters 17-23
            Called from Alcotest_engine__Monad.Identity.catch in file "src/alcotest-engine/monad.ml", line 24, characters 31-35
            
Logs saved to `~/r2c/semgrep/_build/_tests/semgrep-core/OSemgrep Networking U+003E Basic HTML.002.output'.
 ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────


┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ [FAIL]        OSemgrep Networking > Basic JSON                                     0   GET synchronously.                                                                                                                                                │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
[01.22][DEBUG]: GET synchronously
[01.22][DEBUG]: GET https://api.github.com/
[exception] TLS failure: authentication failure: invalid certificate chain
            Raised at Http_helpers.get_async.(fun) in file "libs/networking/http_helpers.ml", line 24, characters 2-692
            Re-raised at Lwt.Miscellaneous.poll in file "src/core/lwt.ml", line 3123, characters 20-29
            Called from Lwt_main.run.run_loop in file "src/unix/lwt_main.ml", line 27, characters 10-20
            Called from Lwt_main.run in file "src/unix/lwt_main.ml", line 106, characters 8-13
            Re-raised at Lwt_main.run in file "src/unix/lwt_main.ml", line 112, characters 4-13
            Called from Unit_Networking.get_and_check in file "src/osemgrep/networking/Unit_Networking.ml", line 28, characters 17-37
            Called from Common.fast_map in file "libs/commons/Common.ml", line 199, characters 16-19
            Called from Unit_Networking.get_and_check_multi in file "src/osemgrep/networking/Unit_Networking.ml", line 45, characters 2-34
            Called from Alcotest_engine__Core.Make.protect_test.(fun) in file "src/alcotest-engine/core.ml", line 181, characters 17-23
            Called from Alcotest_engine__Monad.Identity.catch in file "src/alcotest-engine/monad.ml", line 24, characters 31-35
            
Logs saved to `~/r2c/semgrep/_build/_tests/semgrep-core/OSemgrep Networking U+003E Basic JSON.000.output'.
 ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────


┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ [FAIL]        OSemgrep Networking > Basic JSON                                     1   GET asynchronously.                                                                                                                                               │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
[01.31][DEBUG]: GET asynchronously (parallel)
[01.31][DEBUG]: (lwt) GET https://api.github.com/
[exception] TLS failure: authentication failure: invalid certificate chain
            Raised at Http_helpers.get_async.(fun) in file "libs/networking/http_helpers.ml", line 24, characters 2-692
            Re-raised at Unit_Networking.get_and_check_lwt.(fun) in file "src/osemgrep/networking/Unit_Networking.ml", line 37, characters 2-191
            Re-raised at Unit_Networking.get_and_check_multi_lwt.(fun) in file "src/osemgrep/networking/Unit_Networking.ml", line 53, characters 11-81
            Re-raised at Lwt.Miscellaneous.poll in file "src/core/lwt.ml", line 3123, characters 20-29
            Called from Lwt_main.run.run_loop in file "src/unix/lwt_main.ml", line 27, characters 10-20
            Called from Lwt_main.run in file "src/unix/lwt_main.ml", line 106, characters 8-13
            Re-raised at Lwt_main.run in file "src/unix/lwt_main.ml", line 112, characters 4-13
            Called from Alcotest_engine__Core.Make.protect_test.(fun) in file "src/alcotest-engine/core.ml", line 181, characters 17-23
            Called from Alcotest_engine__Monad.Identity.catch in file "src/alcotest-engine/monad.ml", line 24, characters 31-35
            
Logs saved to `~/r2c/semgrep/_build/_tests/semgrep-core/OSemgrep Networking U+003E Basic JSON.001.output'.
 ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────


┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ [FAIL]        OSemgrep Networking > Basic JSON                                     2   GET asynchronously (parallel).                                                                                                                                    │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
[01.39][DEBUG]: GET asynchronously (parallel)
[01.39][DEBUG]: (lwt) GET https://api.github.com/
[01.39][DEBUG]: (lwt) GET https://dummyjson.com/products/1
[01.39][DEBUG]: (lwt) GET http://ip.jsontest.com
ASSERT Body is not empty
ASSERT Body is not empty
[exception] TLS failure: authentication failure: invalid certificate chain
            Raised at Lwt.Miscellaneous.poll in file "src/core/lwt.ml", line 3123, characters 20-29
            Called from Lwt_main.run.run_loop in file "src/unix/lwt_main.ml", line 27, characters 10-20
            Called from Lwt_main.run in file "src/unix/lwt_main.ml", line 106, characters 8-13
            Re-raised at Lwt_main.run in file "src/unix/lwt_main.ml", line 112, characters 4-13
            Called from Alcotest_engine__Core.Make.protect_test.(fun) in file "src/alcotest-engine/core.ml", line 181, characters 17-23
            Called from Alcotest_engine__Monad.Identity.catch in file "src/alcotest-engine/monad.ml", line 24, characters 31-35
            
Logs saved to `~/r2c/semgrep/_build/_tests/semgrep-core/OSemgrep Networking U+003E Basic JSON.002.output'.
 ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

**Comment by aryx** [comment]
Maybe we should have a separate test target (and GHA workflow) for those kind of tests that can fail because of networking issues are so are more flaky. 
Test.ml should be for really determistic clear simple tests.
We do a little bit the same in pysemgrep with mostly stable e2e tests, and then the parsing tests that clone github repo and parse and are super slow and sometimes fail because of github timeout or other shenanigans.

**Comment by ajbt200128** [comment]
@aryx these should not be flaky tests, but I agree these should probably be somewhere else. The actualy tls/ssl issue is more pressing tho

File: src/osemgrep/networking/Unit_Networking.ml [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  in
+  pack_tests "Basic JSON"
+    [
+      ("GET synchronously", get_sync);
+      ("GET asynchronously", get_async);
+      ("GET asynchronously (parallel)", get_async_parallel);
+    ]
+
+let post_tests () =
+  let url_body_pairs = [ ("https://postman-echo.com/post", "") ] in
```
Comment (zzeleznick) [inline src/osemgrep/networking/Unit_Networking.ml:146]: ty!

**Comment by zzeleznick** [comment]
If possible, I'd like to debug print which `tls_library` we're running with these tests.

Screenshot 2023-09-19 at 3 32 57 PM

https://github.com/dsheets/ocaml-conduit/blob/master/lib/conduit_lwt_unix.ml#L35

On my machine running `CONDUIT_DEBUG=true COHTTP_DEBUG=true CONDUIT_TLS=openssl SSL_CERT_FILE=~/Downloads/ca-certificates.crt cohttp-curl-lwt -v -X GET https://tls-v1-2.badssl.com:1012` will show

```
Selected TLS library: OpenSSL
```

whereas omitting `CONDUIT_TLS` will default to `native` instead of `openssl`

**Review by aryx — APPROVED** [review]

**Pull request merged**